src/browser/encoding/text_encoder.zig

```diff
@@ -0,0 +1,60 @@
+// Copyright (C) 2023-2024  Lightpanda (Selecy SAS)
+//
+// Francis Bouvier <francis@lightpanda.io>
+// Pierre Tachoire <pierre@lightpanda.io>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+
+const SessionState = @import("../env.zig").SessionState;
+
+pub const Interfaces = .{
+    TextEncoder,
+};
+
+// https://encoding.spec.whatwg.org/#interface-textencoder
+pub const TextEncoder = struct {
+    pub fn constructor() !TextEncoder {
+        return .{};
+    }
+
+    pub fn get_encoding(_: *const TextEncoder) []const u8 {
+        return "utf-8";
+    }
+
+    pub fn _encode(_: *const TextEncoder, v: []const u8, state: *SessionState) ![]u8 {
+        // ensure the input is a valid utf-8
+        if (!std.unicode.utf8ValidateSlice(v)) {
+            return error.InvalidUtf8Input;
+        }
+
+        // raw copy b/c we know the input is utf-8
```

Comment:
I don't think you'll need to allocate, and can use `v` directly:

```zig
pub fn _encode(_: *const TextEncoder, v: []const u8) !Env.TypedArray(u8) {
  ....
  return .{.values = v}
}
```

If you _did_ need to allocate, you could use the `call_arena`, which is shorter-lived, since the data gets copied into v8 anyways.